🚀 4단계 - Controller 메서드 인자 매핑

nainu opened the pull request:
안녕하세요, 한솔님. 
난이도가 높아져서 푸는 방법을 고민하다가 엄청 오래 걸렸습니다.
이번에도 잘 부탁드립니다. :)

nainu committed:
docs: README.md (4단계)

nainu committed:
test: FilterTest, ServletTest 러닝테스트

nainu committed:
feat: Controller 메서드 인자 파싱

nainu committed:
feat: ResponseBody 가 있는 경우 JSON 으로 응답

nainu committed:
refactor: ModelAndView 에 validateView 추가

nainu commented:
https://github.com/next-step/java-mvc/pull/32#discussion_r1694257629
요건 여유가 안 나서 일단 반영하지 않았습니다.
당장 생각나는 방법은 DI로 주입해주는 것이었는데 DispatcherServlet 생성하는 MyWebApplicationInitializer 에서 직접 넣어줘야 하고 여기로 옮기는게 당장의 개선이 될 것 같지 않았습니다. (일단 요기까지만 고민)

nainu commented on mvc/src/main/java/com/interface21/webmvc/servlet/ModelAndView.java:
```diff
@@ -12,6 +12,22 @@ public class ModelAndView {
+        validateView(view);
+    }
+
+    private void validateView(Object view) {
+        if (view == null) {
+            return;
+        }
+        if (view instanceof View) {
+            return;
+        }
+        if (view instanceof String) {
+            return;
+        }
+
+        throw new IllegalArgumentException("invalid view: " + view);
     }
```
https://github.com/next-step/java-mvc/pull/32#discussion_r1694257544
일단 여기서 문자열도 View 도 아닌 경우를 제외시켰습니다.
DispatcherServlet 에서도 View/String 이 아닌 경우 예외를 발생시키기 때문에 빈틈은 없을 것 같은데,
이런 로직이 한군데 모여있지 않은 것은 조금 찝찝합니다.
view 가 3가지 타입(1.View 2.String 3.그외) 이 있는 상황이어서, 그 상태를 반환하는 것도 만들어볼까 했다가 일이 커져서 멈췄습니다.
개선해볼만한 방법이 있을까요?

catsbi replied:
요청에 대한 핸들링은 지금 모두 DispatcherServlet에서 이뤄지고 있습니다. 
요청을 처리할 수 있는 적절한 핸들러를 핸들러 일급컬렉션에서 찾은뒤 핸들러를 지원하는 적절한 어댑터를 찾아 어댑터에게 동작을 위임하고 결과를 반환받죠. 

그럼 그 결과를 렌더링하기 위해서 현재는 결과물 자체가 내부적으로 유효성 검증을 수행하고 있는데, 
그 결과물 역시 핸들러와 동일하게 결과물 처리 (ex: viewResolvers) 핸들러가 찾아서 처리하도록 하면 어떨까 싶네요. 
예시를 수도코드로 만들면 다음과 같이 동작할 수 있을 것 같네요
```java
@Override
    protected void service(final HttpServletRequest request,
                           final HttpServletResponse response) {
        HandlerExecution handlerExecution = getHandler(request);

        Object handleResult = process(handlerExecution, request, response);

        ViewResolver foundResolver = viewResolvers.findResolver(handlerResult, request, response);

        foundResolver.render(handlerResult, request, response);
    }

```

nainu replied:
위 코드 참고해서 말씀하신 대로 변경해 보았습니다. view resolving 과 view render 코드가 나눠져 있던 것이 적당히 뭉쳐져서 좀더 응집된 형태가 되었습니다. 좋은 것 같습니다.!! 👍

nainu commented on mvc/src/main/java/com/interface21/web/bind/annotation/ResponseBody.java:
```diff
@@ -0,0 +1,10 @@
+@Target({ElementType.TYPE, ElementType.METHOD})
+@Retention(RetentionPolicy.RUNTIME)
+@Documented
+public @interface ResponseBody {
+
+}
```
ResponseBody 지원을 위해 어노테이션 추가

nainu commented on app/src/main/webapp/register.jsp:
```diff
@@ -16,6 +16,7 @@
                                     <div class="card-header"><h3 class="text-center font-weight-light my-4">회원 가입</h3></div>
                                     <div class="card-body">
                                         <form method="post" action="/register">
+                                            <input name="id" value="2" type="hidden" />
```
임시로.. form 에서 User 를 받을 수 있게 하기 위해서 id도 넣어주었습니다.

```java
    @RequestMapping(value = "/register", method = RequestMethod.POST)
    public ModelAndView save(final User user) {
        InMemoryUserDao.save(user);
        
        return new ModelAndView("redirect:/index.jsp");
    }
```

catsbi replied:
2라는 값을 강제로 하드코딩하기보다는 InMemoryUserDao를 수정하셔도 좋을 것 같네요. 🤔

nainu replied:
여러번 생성하면 새로운 id 를 발급받을 수 있도록 하고, 위 코드는 삭제해 보았습니다. :)

nainu commented on mvc/src/main/java/com/interface21/webmvc/servlet/mvc/tobe/viewresolver/JsonViewResolver.java:
```diff
@@ -1,14 +1,17 @@
     @Override
-    public View resolveView(String viewName) {
+    public View resolveView(String viewName, HandlerExecution handlerExecution) {
+        if (handlerExecution.isResponseBodyAnnotated()) return new JsonView();
```
https://github.com/next-step/java-mvc/pull/32#discussion_r1694257473

ResponseBody 가 달려있는지 여기서 확인하고 JSON 으로 응답하도록 해 보았습니다.

nainu commented on mvc/src/main/java/com/interface21/webmvc/servlet/mvc/tobe/parameterresolver/RequestParameterResolver.java:
```diff
@@ -0,0 +1,94 @@
+import java.util.Arrays;
+import java.util.List;
+
+public class RequestParameterResolver implements ParameterResolver {
```
얘가 좀 로직이 많은데요, HttpServletRequest/Response, HttpSession 제외한 나머지 타입을 다 처리하는 리졸버입니다.

nainu commented on mvc/src/main/java/com/interface21/webmvc/servlet/mvc/tobe/ViewResolver.java:
```diff
@@ -3,5 +3,5 @@
 
 public interface ViewResolver {
-    View resolveView(String viewName);
+    View resolveView(String viewName, HandlerExecution handlerExecution);
```
ResponseBody 어노테이션이 달려있는지 확인하기 위해서 HandlerExecution 을 던져주었습니다.

nainu commented on mvc/src/main/java/com/interface21/webmvc/servlet/mvc/tobe/HandlerExecution.java:
```diff
@@ -1,22 +1,62 @@
 
 public class HandlerExecution {
 
     private final Object target;
     private final Method method;
+    private final HandlerKey handlerKey;
+    private List<ParameterResolver> parameterResolvers;
 
-    public HandlerExecution(Object target, Method method) {
+    public HandlerExecution(Object target, Method method, HandlerKey handlerKey) {
         this.target = target;
         this.method = method;
+        this.handlerKey = handlerKey;
+
+        initParameterResolvers();
+    }
+
+    private void initParameterResolvers() {
+        this.parameterResolvers = new ArrayList<>();
+        this.parameterResolvers.add(new HttpServletRequestResolver());
+        this.parameterResolvers.add(new HttpServletResponseResolver());
+        this.parameterResolvers.add(new HttpSessionResolver());
+        this.parameterResolvers.add(new RequestParameterResolver(handlerKey));
     }
 
     public ModelAndView handle(final HttpServletRequest request, final HttpServletResponse response) throws Exception {
-        return (ModelAndView) method.invoke(target, request, response);
+        Object[] resolvedArguments = Arrays.stream(method.getParameters())
+                                           .map(parameter -> resolveParameter(parameter, request, response))
+                                           .toArray();
+
+        return (ModelAndView) method.invoke(target, resolvedArguments);
+    }
```
컨트롤러 메서드의 인자를 변환해주는 걸 요기서 시작합니다.

catsbi commented on mvc/src/main/java/com/interface21/webmvc/servlet/mvc/tobe/AnnotationHandlerMapping.java:
```diff
@@ -67,19 +61,23 @@ private void registerMappings(Method methodType, Map<Class<?>, Object> controlle
-                    new HandlerExecution(controllerObject, methodType)
-            );
+            HandlerKey handlerKey = new HandlerKey(uri, requestMethod);
+            handlerExecutions.put(handlerKey, new HandlerExecution(controllerObject, methodType, handlerKey));
```
handlerKey가 동일한케이스가 발생하면 어떻게될까요? 그냥 덮어씌우면 어떻게 될까요?

nainu replied:
실제로 spring boot 프로젝트를 하나 만들어 테스트해봤더니 예외가 발생하더라구요. 새로운 예외케이스를 알게 되어 의미가 있었습니다. :+1: `new IllegalStateException("이미 동일한 매핑이 등록되어 있습니다.")` 발생시키도록 수정했습니다.

catsbi commented on mvc/src/main/java/com/interface21/webmvc/servlet/mvc/tobe/DispatcherServlet.java:
```diff
@@ -2,23 +2,20 @@
-import java.util.Collections;
-import java.util.List;
-import java.util.Objects;
+import java.util.*;
```
import wildcard는 권장드리지 않아요. 소스코드에서 사용되는 모듈들을 모두 묶어버려서 내가 사용하는 모듈이 무엇인지 파악하기 힘들게 하는 단점이 있어서요 ㅎㅎ

nainu replied:
일괄 제거하였습니다. 예리한 리뷰 감사합니다. :)

catsbi commented on mvc/src/main/java/com/interface21/webmvc/servlet/mvc/tobe/DispatcherServlet.java:
```diff
@@ -54,74 +51,87 @@ private void initViewResolvers() {
+                                                 .map(handlerMapping -> handlerMapping.getHandler(request))
+                                                 .filter(Objects::nonNull)
+                                                 .findFirst()
+                                                 .orElse(null);
```
null을 반환하면 process에서 NPE가 발생하지 않을까요?

nainu replied:
뒤쪽 메서드에서 null check 를 하고 있긴 했었는데요. 깔끔한 예외처리가 아닌 듯 해서 not null 객체들을 반환하도록 근처 코드들을 수정해 보았습니다.

catsbi commented on mvc/src/main/java/com/interface21/webmvc/servlet/mvc/tobe/HandlerExecution.java:
```diff
@@ -1,22 +1,62 @@
+        initParameterResolvers();
+    }
+
+    private void initParameterResolvers() {
```
HandlerExecution은 경로 갯수에 따라 무수히 많이 생성될 객체인데 각 객체마다 parameterResolvers가 각각 생성되어 관리되는게 적절할까요?

nainu replied:
맞습니다. 당장 동작하는 것으로 PR 을 올렸었는데요, 분리 해 내면서 좀더 코드 정리가 되었습니다. :+1:

분리하려고 보니 handlerKey 가 전달되는 코드 때문에 뜯어내는 데에 시간이 조금 걸렸는데요. ParameterResolver 인터페이스의 resolve 메서드에 HandlerKey handlerKey 를 함께 인자로 넘겨주는 것으로 일단 해소하였습니다.

catsbi commented on mvc/src/main/java/com/interface21/webmvc/servlet/mvc/tobe/HandlerExecution.java:
```diff
@@ -1,22 +1,62 @@
+                                 .orElseThrow(() -> new NoSuchElementException("No value present") /*XXX: 예외 정의할 것*/);
+    }
+
+    public boolean isResponseBodyAnnotated() {
```
HandlerExecution이라는 객체에서 ReponseBody라는 애노테이션하나에 의존하게 되는데 차라리 hasAnnotated(Class<T> annotationClass) 를 만들어서 범용적으로 쓸 수 있게 하는건 어떨까요

nainu replied:
반영했습니다.! 당장 다른곳에서 쓰는 곳은 아직 없지만 여기저기 쓰일 수도 있을법 해 보입니다. :+1:

catsbi commented on mvc/src/main/java/com/interface21/webmvc/servlet/mvc/tobe/parameterresolver/RequestParameterResolver.java:
```diff
@@ -0,0 +1,94 @@
+import java.util.List;
+
+public class RequestParameterResolver implements ParameterResolver {
+    private static final List<Type> SIMPLE_TYPES = List.of(
```
Primitive와 Wrapper 타입은 이 밖에도 boolean, Boolean, byte, Byte, char, Character, double, Double등 많을 것 같네요!

nainu replied:
일단 가능해보이는 primitive type 들을 배열에 집어넣고 구현했습니다. 테스트 코드 `AnnotationHandlerMappingTest` 에 케이스 추가해서 각 타입들 resolve 되는지 테스트하였습니다.

catsbi commented on mvc/src/main/java/com/interface21/webmvc/servlet/mvc/tobe/parameterresolver/RequestParameterResolver.java:
```diff
@@ -0,0 +1,94 @@
+        }
+        if (parameterType.equals(long.class)) {
+            return Long.parseLong(value);
+        }
```
위에서 선언한 타입에 비해서 여기서 변환에 대응하는건 primitive타입뿐이고 int, long 두 타입뿐이네요! 다른 타입들로 선언하면 에러가 날 것 같아요!

nainu replied:
요 타입들도 더 확장해서 위 primitive 타입들을 박싱한 타입들은 전부 동작하도록 수정했습니다!

catsbi commented on mvc/src/main/java/com/interface21/webmvc/servlet/mvc/tobe/HandlerExecution.java:
```diff
@@ -1,22 +1,62 @@
+        this.parameterResolvers.add(new HttpServletRequestResolver());
+        this.parameterResolvers.add(new HttpServletResponseResolver());
+        this.parameterResolvers.add(new HttpSessionResolver());
+        this.parameterResolvers.add(new RequestParameterResolver(handlerKey));
```
RequestParam뿐 아니라 ModelAttribute 혹은 객체 타입에도 대응가능한 ParameterResolver가 필요해보이네요. 요구사항에도 객체 타입이 있었구요 ㅎㅎ

nainu replied:
첨에 PR 올릴땐 파라메터 타입만 보고 단순하게 처리했다면 지금은 조금 더 여러 ParameterResolver 로 분리해서 처리되도록 수정해 보았습니다.
- HttpServletRequestResolver - 그대로 (HttpServletRequest 타입 처리)
- HttpServletResponseResolver - 그대로 (HttpServletResponse 타입 처리)
- HttpSessionResolver - 그대로 (HttpSession 타입 처리)
- PathVariableParameterResolver - @PathVariable 달려있는 파라메터만 처리
- RequestParameterResolver - primitive 타입, 그걸 박싱한 타입들만 처리
- StructuredRequestParameterResolver - 클래스 형태로 돼 있는 타입들 처리

찾아보니 아마 ModelAttribute 가 바로 위 StructuredRequestParameterResolver 에서 처리하고 있는 형태의 파라메터에 달아주는 어노테이션인 것 같았습니다. 매번 생략하고 썼던 것 같은데, 작업하면서 여기저기 문서를 읽어볼 기회가 있었던 것 같습니다. :+1:

catsbi commented on mvc/src/test/java/com/interface21/webmvc/servlet/mvc/tobe/AnnotationHandlerMappingTest.java:
```diff
@@ -50,21 +51,92 @@ void post() throws Exception {
     }
+
+    @Test
+    void testUsersCreateString() throws Exception {
```
테스트를 작성할 때 적절한 주석(Ex: DisplayName)이 없을 경우 테스트가 수 백개가 넘어가는 시점에서는 사실상 파악이 불가능하고 그럼 유지보수가 안될 가능성이 대부분입니다. 

코드 변경이 크게 일어나서 테스트가 관리되야할 때 나중에 들어온 개발자가 테스트를 파악하려면 시간이 부족하고, 그러다보면 그냥 주석처리되거나 삭제될 수가 있거든요. 

그래서 테스트코드는 더욱 더 유지보수성이 중요합니다 👍

nainu replied:
감사합니다. 설명 추가해 주었습니다. :+1:

catsbi reviewed (changes requested):
동윤님 어느덧 마지막 4단계 미션이네요! 고생하셨습니다 👏 
구조 변경이 꽤 많은 것 같네요 ㅎㅎ 
잘 구현해주셨는데, 몇 가지 보완하면 좋을 부분들이 있어서 코멘트 남겨드렸으니 확인 후 다시 리뷰요청 부탁드려요!

nainu committed:
refactor: ParameterResolver 하위클래스 구현 정리

- ParameterResolver 정의를 AnnotationHandlerMapping 으로 분리
- 리졸브 할 수 없는 타입의 경우 NotSupportedTypeResolveException

nainu committed:
refactor: HandlerExecution.hasAnnotated(Annotation) 도입

nainu committed:
refactor: import 의 wildcard 제거

nainu committed:
refactor: DispatcherServlet 의 getHandler(request) 가 null를 리턴하지 않음

nainu committed:
fix: 동일한 매핑이 발견될 경우 IllegalStateException 발생

nainu committed:
refactor: ViewResolvers 도입, 렌더링 로직을 ViewResolver로 넘김

nainu committed:
fix: 가입시 user id 도 자동 생성되도록 수정

nainu commented:
리뷰 넘 잘해주셔서 아예 새로 생각하게 된 부분도 있어서 의미있게 긴(ㅋㅋ)시간을 보냈습니다.
너무 텀이 길게 작업하게 되었는데, 다시 한번 리뷰 요청드립니다.

감사합니다. :)

catsbi commented on app/src/main/java/camp/nextstep/dao/InMemoryUserDao.java:
```diff
@@ -15,7 +15,19 @@ public class InMemoryUserDao {
+        database.put(newUser.getAccount(), newUser);
+    }
+
+    private static long getNewId() {
```
매번 모든 아이디를 정렬해서 최대값을 구해야하기보다는 static하게 관리되는 시퀀스를 하나 만들어서 관리해도 좋을 것 같네요 ㅎㅎ

catsbi commented on mvc/src/main/java/com/interface21/core/util/ConversionUtil.java:
```diff
@@ -0,0 +1,40 @@
+
+    public static Object boxPrimitiveValue(String value, Type parameterType) {
+        if (parameterType.equals(int.class)) {
+            return value == null ? 0 : Integer.parseInt(value);
```
Java에서는 primitive type으로 변환할 때 null일 경우 MissingServletRequestParameterException 예외가 던져집니다. 
Integer와 같은 Wrapper 타입일 경우에는 null로 주입되구요. 

하지만, 이 함수를 호출했을땐 null이여도 기본값을 주입해주게 되는데요. RequestParam애노테이션에서 default를 선언했을때와 그러지 않았을 경우에 로직이 돌 수 있는부분도 같이 고민해보면 좋을 것 같아요 ㅎㅎ

catsbi commented on mvc/src/main/java/com/interface21/core/util/ConversionUtil.java:
```diff
@@ -0,0 +1,40 @@
+            return value.charAt(0);
+        }
+        if (parameterType.equals(char[].class)) {
+            return value == null ? EMPTY_CHARS : value.toCharArray();
```
자바에서 삼항연산자는 사실 if-else로 권장되는 패턴은 아니긴합니다 ㅎㅎ

catsbi commented on mvc/src/main/java/com/interface21/webmvc/servlet/mvc/tobe/AnnotationHandlerMapping.java:
```diff
@@ -47,6 +57,15 @@ private void registerRequestMapping() {
         }
     }
 
+    private void registerParameterResolvers() {
```
개인적으로 WebConfig같은 계층이 하나 나오면 좋지 않을까 생각이 드네요. 지금은 각 객체에서 어떤 구현체를 의존하고 추가하는지를 보려면 모두 직접 찾아가봐야하니 DIP 원칙도 위배되고 추적 용이성도 떨어지는 느낌이에요 ㅎㅎ

catsbi commented on mvc/src/main/java/com/interface21/webmvc/servlet/mvc/tobe/DispatcherServlet.java:
```diff
@@ -1,40 +1,37 @@
 import org.slf4j.LoggerFactory;
 
-import java.util.*;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
```
모든 와일드카드를 다 풀어주셨군요 👍

catsbi commented on mvc/src/main/java/com/interface21/webmvc/servlet/mvc/tobe/ViewResolvers.java:
```diff
@@ -0,0 +1,37 @@
+
+    public ViewResolver findResolver(@Nullable ModelAndView modelAndView,
+                                     @Nullable HandlerExecution handlerExecution) {
+        if (modelAndView == null) {
+            throw new IllegalStateException("템플릿을 찾을 수 없습니다");
+        }
```
IllegalStateException은 클라이언트가 어떤 값을 줘도 현재 상황에선 문제가 되는 경우 발생하는 예외로 이 경우에는 ModelAndView만 정상이면 해결될 문제라 IllegalArgumentException이 더 적절하지 않을까요 🤔

catsbi commented on mvc/src/main/java/com/interface21/webmvc/servlet/mvc/tobe/parameterresolver/StructuredRequestParameterParameterResolver.java:
```diff
@@ -0,0 +1,59 @@
+import java.lang.reflect.Parameter;
+import java.util.Arrays;
+
+public class StructuredRequestParameterParameterResolver implements ParameterResolver {
```
객체 타입의 리졸버를 잘 만들어주셨네요 👍

catsbi reviewed (approved):
동윤님, 피드백 계속 반영 열심히 해주셔서 감사합니다 👍 
마지막인만큼 좀 더 인사이트 얻기를 바라는 마음에 열심히 드렸던 것 같네요 ㅎㅎ

좀 더 고민해보면 좋을 코멘트 몇개 더 남겨드렸는데, 크게 문제될정도는 아닌 부분들이라 추후에 고민해보시는걸로하고 이번 미션은 여기서 마무리하도록 하겠습니다. 고생 많으셨습니다!
